odxtools/statetransitionref.py

```diff
@@ -0,0 +1,219 @@
+        else:
+            tr = resolve_snref(inner_param_value[0], inner_param.table.table_rows)
+
+        if tr.structure is None:
+            odxraise("Table row must reference a structure")
+            return None, None
+
+        inner_param_list = tr.structure.parameters
+        inner_param_value = inner_param_value[1]
+    else:
+        dop = getattr(inner_param, "dop", None)
+        if isinstance(dop, BasicStructure):
+            inner_param_list = dop.parameters
+        else:
+            odxraise(f"Expected the referenced parameter to be composite")
+
+    if not isinstance(inner_param_value, dict):
+        odxraise(f"Expected the referenced parameter to be composite")
+        return None, None
+
+    return _resolve_in_param_helper(inner_param_list, inner_param_value, path_chunks[1:])
+
+
+@dataclass
+class StateTransitionRef(OdxLinkRef):
+    """Describes a state transition that is to be potentially taken if
+    a diagnostic communication is executed
+
+    Besides the "raw" state transistion, state transition references
+    may also be conditional on the observed response of the ECU.
+
+    """
+    value: Optional[str]
+
+    in_param_if_snref: Optional[str]
+    in_param_if_snpathref: Optional[str]
+
+    @property
+    def state_transition(self) -> StateTransition:
+        return self._state_transition
+
+    @staticmethod
+    def from_et(  # type: ignore[override]
+            et_element: ElementTree.Element,
+            doc_frags: List[OdxDocFragment]) -> "StateTransitionRef":
+        kwargs = dataclass_fields_asdict(OdxLinkRef.from_et(et_element, doc_frags))
+
+        value = et_element.findtext("VALUE")
+
+        in_param_if_snref = None
+        if (in_param_if_snref_elem := et_element.find("IN-PARAM-IF-SNREF")) is not None:
+            in_param_if_snref = odxrequire(in_param_if_snref_elem.get("SHORT-NAME"))
+
+        in_param_if_snpathref = None
+        if (in_param_if_snpathref_elem := et_element.find("IN-PARAM-IF-SNPATHREF")) is not None:
+            in_param_if_snpathref = odxrequire(in_param_if_snpathref_elem.get("SHORT-NAME-PATH"))
+
+        return StateTransitionRef(
+            value=value,
+            in_param_if_snref=in_param_if_snref,
+            in_param_if_snpathref=in_param_if_snpathref,
+            **kwargs)
+
+    def __post_init__(self) -> None:
+        if self.value is not None:
+            odxassert(self.in_param_if_snref is not None or self.in_param_if_snref is not None,
+                      "If VALUE is specified, a parameter must be referenced")
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        return {}
+
+    def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+        self._state_transition = odxlinks.resolve(self, StateTransition)
+
+    def _resolve_snrefs(self, context: SnRefContext) -> None:
+        pass
+
+    def execute(self, state_machine: StateMachine, params: List[Parameter],
```

Comment:
Looks like same code from `applies`

Reply:
true. the bones are the same, but the meat attached to them is slightly different. I did not see a simple way to unify these two...

Reply:
No callbacks needed

` def get_next_state(self, state_machine: StateMachine, params,  param_value_dict) -> Optional[State]:`

Reply:
for the precondition state references, there is no "next state", though.

Reply:
okay, I've unified that boiler plate code: https://github.com/mercedes-benz/odxtools/pull/398/commits/93e57cc0c3702911e6f4e634cdaab1b1b5ff61ed . do you see any further holdups for getting this merged?